dnprogs/contrib/ph3-der-loewe/man1/dnetstat.1
.TH "dnetstat" "1" "September 2008" "dntools" "System Manager's Manual"

.SH NAME

dnetstat \- lists DECnet connections

.SH SYNOPSIS

dnetstat [\-n] [\-h|\-\-help]

.SH "DESCRIPTION"
\fBdnetstat\fR is a \fBnetstat\fR(8) like tool that displays current DECnet network connections.

.SH OPTIONS
.SS "\-n"
Show nodenames and numbered objects in as numerical addresses and don't resolve names.
.SS "\-\-help, \-h"
Prints a small help.

.SH SEE ALSO
\fBnetstat\fB(8).

.SH AUTHORS
This program is written by Philipp "ph3-der-loewe" Schafft <lion@lion.leolix.org>.


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/contrib/ph3-der-loewe/dnetstat.c
//dnetstat.c:

/*
    copyright 2008 Philipp 'ph3-der-loewe' Schafft <lion@lion.leolix.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    version 3.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*/

#include <stdio.h>
#include <string.h>
#include <errno.h>

#include <unistd.h>
#include <stdlib.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <fcntl.h>

#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>

#include <netdb.h>

#define DNNS_FILE "/proc/net/decnet"

struct dn_nse {
 char node[8];
 char object[32];
};

char * progname = NULL;
int numeric = 0;

void usage (void) {
 fprintf(stderr, "Usage: %s [OPTIONS]\n", progname);
 fprintf(stderr, "\n");
 fprintf(stderr, "Options:\n");
 fprintf(stderr, " -h   --help        this help\n"
                 " -n                 numerical mode (do not show node names but addresses)\n"
        );
}

char * object_name(char *number) {
 int objnum = atoi(number);
 static char name[16];

 if (numeric)
  return number;

 if ( objnum )
  if ( getobjectbynumber(objnum, name, 16) != -1 )
   return name;

 return number;
}

int prep_addr (char * buf, char * object) {
 struct nodeent * ne;

 *index(buf, '/') = 0;

  if ( strcmp(buf, "0.0") == 0 && ! numeric ) {
   strcpy(buf, "*");
  } else if ( !numeric ) {
   if ( (ne = getnodebyname(buf)) != NULL ) {
    if ( (ne = getnodebyaddr((const char *)ne->n_addr, ne->n_length, ne->n_addrtype)) != NULL ) {
     strcpy(buf, ne->n_name);
    }
   }
  }

 strcat(buf, "::");

 if ( strcmp(object, "0") == 0 && ! numeric ) {
  strcat(buf, "*");
 } else {
  strcat(buf, object_name(object));
 }
 return 0;
}

char * state_ktou (char * state, char ** dir) {

 if ( strcmp(state, "OPEN") == 0 ) {
  *dir = "IN";
  return "LISTEN";
 } else if ( strcmp(state, "RUN") == 0 ) {
  return "ESTABLISHED";
 } else if ( strcmp(state, "RJ") == 0 ) {
  *dir = "OUT";
  return "REJECTED";
 } else if ( strcmp(state, "DN") == 0 ) {
  return "DISCNOTIFY";
 } else if ( strcmp(state, "DIC") == 0 ) {
  return "DISCONNECTED";
 } else if ( strcmp(state, "DI") == 0 ) {
  return "DISCONNECTING";
 } else if ( strcmp(state, "DR") == 0 ) {
  return "DISCONREJECT";
 } else if ( strcmp(state, "DRC") == 0 ) {
  return "DISCONREJCOMP";
 } else if ( strcmp(state, "CL") == 0 ) {
  return "CLOSED";
 } else if ( strcmp(state, "CN") == 0 ) {
  return "CLOSEDNOTIFY";
 } else if ( strcmp(state, "CI") == 0 ) {
  *dir = "OUT";
  return "CONNECTING";
 } else if ( strcmp(state, "NR") == 0 ) {
  return "NORES";
 } else if ( strcmp(state, "NC") == 0 ) {
  return "NOCOM";
 } else if ( strcmp(state, "CC") == 0 ) {
  return "CONCONFIRM";
 } else if ( strcmp(state, "CD") == 0 ) {
  return "CONDELIVERY";
 } else if ( strcmp(state, "CR") == 0 ) {
  return "CONNECTRECV";
 }

 return state;
}

int proc_file (FILE * fh) {
 struct dn_nse local, remote;
 char state[32], immed[32];
 char buf[1024];
 char out[1024] = {0}, * outdir = out+57;
 char conid[8] = {0,0,0,0,0,0,0,0}, *lid, *rid;
 char * lbuf = buf, * rbuf = buf + 512;
 char * dir = ""; // max be "UNI" or "BI", "IN", "OUT"
 int unused;

 if ( fgets(buf, 1024, fh) == NULL ) {
  fprintf(stderr, "Error: can not read banner from file\n");
  return -1;
 }

 if ( strcmp(buf, "Local                                              Remote\n") != 0 ) {
  fprintf(stderr, "Error: invalid file format\n");
  return -1;
 }

 while (fscanf(fh, "%s %04d:%04d %04d:%04d %01d %16s"
                   "%s %04d:%04d %04d:%04d %01d %16s"
                   "%4s %s\n",
                   lbuf, &unused, &unused, &unused, &unused, &unused, local.object,
                   rbuf, &unused, &unused, &unused, &unused, &unused, remote.object,
                   state, immed
                   ) == 16) {
  lid = index(lbuf, '/') + 1;
  rid = index(rbuf, '/') + 1;

  if ( memcmp(lid, conid+4, 4) == 0 && memcmp(rid, conid, 4) == 0 && strcmp(state, "OPEN") != 0 ) {
   if ( *out ) {
    memcpy(outdir, "LOC", 3);
    puts(out);
    *out = 0;
    continue;
   }
  }

  if ( *out )
   puts(out);

  memcpy(conid,   lid, 4);
  memcpy(conid+4, rid, 4);


  prep_addr(lbuf, local.object);
  prep_addr(rbuf, remote.object);

  dir = "";

  if ( strcmp(state, "OPEN") != 0 ) {
   immed[0] = 0;
  }

  sprintf(out, "decnet %-24s %-24s %-3s %-13s %s", lbuf, rbuf, dir, state_ktou(state, &dir), immed);
 }
 puts(out);

 return 0;
}

int main (int argc, char * argv[]) {
 FILE * fh = NULL;
 int i;
 char * k;
 char * file = DNNS_FILE;

 progname = argv[0];

 for (i = 1; i < argc; i++) {
  k = argv[i];
  if ( strcmp(k, "-n") == 0 ) {
   numeric = 1;
  } else if ( strcmp(k, "-h") == 0 || strcmp(k, "--help") == 0 ) {
   usage();
   return 0;
  } else {
   fprintf(stderr, "Error: unknown parameter %s\n", k);
   usage();
   return 1;
  }
 }

 if ( (fh = fopen(file, "r")) == NULL ) {
  fprintf(stderr, "Error: can not open DECnet netstat file: %s: %s\n", file, strerror(errno));
  return 2;
 }

//      |Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
//      |decnet *::29                    *::0                         LISTEN       IMMED
 printf("Active DECnet sockets (servers and established)\n");
 printf("Proto  Local Address            Foreign Address          Dir State         Accept mode\n");
 proc_file(fh);

 fclose(fh);

 return 0;
}

//ll
